src/Downloader.ts

```diff
@@ -169,52 +176,69 @@ class Downloader {
     }
   }
 
-  public async setBaseUrls(forceRender = null) {
+  private getUrlDirector(capabilitiesList): WikimediaDesktopURLDirector | WikimediaMobileURLDirector | VisualEditorURLDirector {
+    for (const capabilityInfo of capabilitiesList) {
+      if (capabilityInfo.condition) {
+        return new capabilityInfo.Director(capabilityInfo.value)
+      }
+    }
+    throw new Error('No suitable URL director found.')
+  }
+
+  public async setBaseUrlsDirectors(forceRender = null) {
     if (!forceRender) {
       //* Objects order in array matters!
-      this.baseUrl = basicURLDirector.buildDownloaderBaseUrl([
-        { condition: await MediaWiki.hasWikimediaMobileApi(), value: MediaWiki.WikimediaMobileApiUrl.href },
-        { condition: await MediaWiki.hasWikimediaDesktopApi(), value: MediaWiki.WikimediaDesktopApiUrl.href },
-        { condition: await MediaWiki.hasVisualEditorApi(), value: MediaWiki.visualEditorApiUrl.href },
-      ])
+      const articlesCapabilitiesList = [
+        { condition: await MediaWiki.hasWikimediaMobileApi(), value: MediaWiki.WikimediaMobileApiUrl.href, Director: WikimediaMobileURLDirector },
+        { condition: await MediaWiki.hasWikimediaDesktopApi(), value: MediaWiki.WikimediaDesktopApiUrl.href, Director: WikimediaDesktopURLDirector },
+        { condition: await MediaWiki.hasVisualEditorApi(), value: MediaWiki.VisualEditorApiUrl.href, Director: VisualEditorURLDirector },
+      ]
+
+      this.baseUrl = basicURLDirector.buildDownloaderBaseUrl(articlesCapabilitiesList)
+      this.articleUrlDirector = this.getUrlDirector(articlesCapabilitiesList)
```

Comment:
Why the condition does not returns directly the director.... like it was returning directly the baseURL in the past? Concept of capabilites (and related functions) seem useless.

Reply:
I refactored this to `downloader.setUrlsDirectors(mainPageRenderer, articlesRenderer)`